File: cpp/15.cpp
#include <iostream>
using namespace std;

long double ft_factorial(int n)
{
	long double fact;

	fact = 1;
	while (n > 0)
		fact *= n--;
	return (fact);
}

unsigned long long int ft_combination(int n, int r)
{

	return ((unsigned long long int )((ft_factorial(n) / ft_factorial(r)) / ft_factorial(n - r)));
}

int main()
{
	cout << (ft_combination(40,20)) << endl;
	return (0);
}



// Solution using recursive backtracking and bruteforce
/*
typedef struct node
{
	int right;
	int down;
} node;

node **ft_create_lattice(void);
void ft_travel(node **, int, int);

unsigned long int ctr;
const int sides = 20;

int main()
{
	node **lattice;

	lattice = ft_create_lattice();
	ft_travel(lattice,0,0);
	cout << ctr << endl;

	return (0);
}

node **ft_create_lattice()
{
	node **lattice;
	int size;

	size = sides;
	lattice = (node **)(malloc(sizeof(node *) * (size + 1)));
	for (int i = 0; i < size + 1; i++)
	{
		lattice[i] = (node *)(malloc(sizeof(node) * (size + 1)));
	}

	for (int i = 0;i < size; i++)
	{
		for (int j = 0;j < size; j++)
		{
			lattice[i][j].down = i + 1;
			lattice[i][j].right = j + 1;
		}
		// rightmost column
		lattice[i][size].down = i + 1;
		lattice[i][size].right = -1;

		// bottommost row
		lattice[size][i].down = -1;
		lattice[size][i].right = i + 1;
	}

	// bottom right element
	lattice[size][size].down = -1;
	lattice[size][size].right = -1;
	
	return lattice;

}

void ft_travel(node **lattice, int i, int j)
{
	int size;

	size = sides;
	if ((i == size) && (j < size))
		ft_travel(lattice,i,lattice[i][j].right);
	else if ((i < size) && (j == size))
		ft_travel(lattice,lattice[i][j].down,j);
	else if ((i == size) && (j == size))
		{
			ctr++;
			cout << ctr << endl;
			return;
		}
	else
	{
		for (int x = 0; x < 2; x++)
			if ((x == 0) && (lattice[i][j].right != -1))
				ft_travel(lattice,i,lattice[i][j].right);
			else if ((x == 1) && (lattice[i][j].down != -1))
				ft_travel(lattice,lattice[i][j].down,j);	
	}

}
*/

File: Python3/15.py
#!/usr/bin/python3

def ft_factorial(n):
	if n == 1:
		return (1)
	return (n * ft_factorial(n - 1))

def ft_combination(n,r):
	return (ft_factorial(n) // (ft_factorial(r) * ft_factorial(n - r)))

print(ft_combination(40,20))

# Solution using recursive backtracking and bruteforce
'''
def travel(lattice, i, j):
	# _ = input()
	global ctr
	global size
	if i == size and j < size:
		# print('right',end=' ')
		travel(lattice,i,lattice[i][j]['right'])
	elif i < size and j == size:
		# print('down',end=' ')
		travel(lattice,lattice[i][j]['down'],j)
	elif i == size and j == size:
		# print('destination')
		ctr += 1
		print(ctr)
		return
	else:
		for move in moves:
			# print(move,end=' ')
			if move == 'right' and not lattice[i][j]['right'] == None:
				travel(lattice,i,lattice[i][j]['right'])
			elif move == 'down' and not lattice[i][j]['down'] == None:
				travel(lattice,lattice[i][j]['down'],j)	
				

size = 20
lattice = [[None for _ in range(size+1)] for _ in range(size+1)]
moves = ['right','down']
ctr = 0
for i in range(size):
	for j in range(size):
		lattice[i][j] = {"right":j + 1,'down': i + 1}
	lattice[i][size] = {'right':None,'down': i + 1}
	lattice[size][i] = {'right':i+1,'down': None}
lattice[size][size] = {'right':None,'down': None}

travel(lattice,0,0)
print(ctr)
'''